pkg/controller/azure/compute/aks.go

```diff
@@ -0,0 +1,460 @@
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := azure.NewClientCredentialsConfig(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return r.NewAKSClientset(config)
+}
+
+// _createApp - create AD Application with Service Principal
+func (r *Reconciler) _createApp(instance *azurecomputev1alpha1.AKSCluster, client azure.AKSClientsetAPI) (string, string, error) {
+	// check for existing application id - if exist: delete and recreate
+	appID := instance.Status.ApplicationObjectID
+	if appID != "" {
+		if err := client.DeleteApplication(ctx, appID); err != nil && !azure.IsErrorNotFound(err) {
+			return "", "", err
+		}
+	}
+
+	// Generate new url for app using dns prefix and location
+	url, err := applicationURL(instance.Spec.DNSNamePrefix, instance.Spec.Location)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Generate new application password and save it as secret
+	password := azure.NewPasswordCredential("rbac")
+	if _, err := util.ApplySecret(r.kubeclient, applicationSecret(instance, []byte(*password.Value))); err != nil {
+		return "", "", err
+	}
+
+	// Generate application name and create application
+	name := fmt.Sprintf("%s%s", clusterNamePrefix, instance.UID)
+	app, err := client.CreateApplication(ctx, name, url, password)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Create Service Principal
+	_, err = client.CreateServicePrincipal(ctx, *app.AppID)
+	if err != nil {
+		return "", "", err
+	}
+
+	instance.Status.ApplicationObjectID = *app.ObjectID
+
+	// return back AD Application ID, and password
+	return *app.AppID, *password.Value, nil
+}
+
+// reasonCreateClusterFailure to track failed create cluster calls
+const reasonCreateClusterFailure = "Failed to create new cluster"
+
+// _create new AKS cluster and required artifacts (AD App and SP)
+func (r *Reconciler) _create(instance *azurecomputev1alpha1.AKSCluster, client azure.AKSClientsetAPI) (reconcile.Result, error) {
+	// create AKS Cluster
+	clusterName := strings.TrimSuffix(util.TrimToSize(fmt.Sprintf("%s%s", clusterNamePrefix, instance.UID), azure.MaxClusterNameLength), "-")
+
+	// create AD Application
+	appID, appPassword, err := r.createApp(instance, client)
+	if err != nil {
+		return r.fail(instance, reasonCreateClusterFailure, err.Error())
+	}
+
+	future, err := client.CreateCluster(ctx, clusterName, appID, appPassword, instance.Spec)
+	if err != nil {
+		if azure.IsErrorBadRequest(err) {
+			instance.Status.SetFailed(reasonCreateClusterFailure, err.Error())
+			// do not requeue on bad requests
+			return resultDone, r.Update(ctx, instance)
+		}
+		return r.fail(instance, reasonCreateClusterFailure, err.Error())
+	}
+
+	instance.Status.UnsetAllConditions()
+	instance.Status.SetCreating()
+	instance.Status.ClusterName = clusterName
+	instance.Status.State = azurecomputev1alpha1.ClusterStateCreating
+	instance.Status.RunningOperation = future
+	return resultDone, r.Update(ctx, instance)
```

Comment:
TBH, I never fully understood that retry loop. Moreover, I suspect it is probably not really needed if we follow a better hygiene in terms of `Reconcile.requeue`. I ran this PR back-to-back several times and could not reproduce the issue we've seen before. Also note, that the issue was primarily (or exclusively) impacting only `AKS` and `MySQLServer` (both azure) controllers. 

![image](https://user-images.githubusercontent.com/324803/50499954-8dfea900-0a02-11e9-9eec-dfaba65b3143.png)

Reply:
i think the main reason we noticed this behavior for the Azure resources is because they are the only ones that have an overloaded CreateOrUpdate function.  Other providers, when running into similar behavior, would just fail the 2nd Create attempt and subsequent reconciliations would converge on everything being OK, so we wouldn't even notice it.  I am hesitant to just remove this retry loop completely because we'll land right back into the scenario of sending duplicate CreateOrUpdate requests to AKS.  At the least, could we add a sensible "requeue after" value here for when we've successfully made the create request? that would significantly reduce the possibility of this duplicate request bug coming back.

Reply:
Ah, now I see your new logic for first checking if the cluster already exists inside `create`.  That may be sufficient to deal with this then, as long as Azure API is consistent (i.e. returns a value indicating the cluster "exists" immediately after the create request has been accepted).

what do you think still about doing a "requeue after" once we've first made the create call?

Reply:
To prevent a "duplicate" cluster create even I have added: